File: src/data/codes/decision-tree.m
clc;
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'student_performance_data.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(4);
    n = 120;

    Attendance = randi([0 2], n, 1);         % 0=Low,1=Medium,2=High
    Internal_Marks = randi([0 2], n, 1);     % 0=Low,1=Medium,2=High
    Assignment_Status = randi([0 1], n, 1);  % 0=No,1=Yes
    Study_Hours = randi([0 2], n, 1);        % 0=Low,1=Medium,2=High

    Result = strings(n,1);
    for i = 1:n
        score = Attendance(i) + Internal_Marks(i) + Assignment_Status(i) + Study_Hours(i);
        if score >= 4
            Result(i) = "Pass";
        else
            Result(i) = "Fail";
        end
    end

    data = table(Attendance, Internal_Marks, Assignment_Status, Study_Hours, Result);
else
    data = readtable(csvFile);
    % Expected columns:
    % Attendance, Internal_Marks, Assignment_Status, Study_Hours, Result
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

% Convert text categories if present
vars = {'Attendance','Internal_Marks','Assignment_Status','Study_Hours'};
for i = 1:length(vars)
    if iscellstr(data.(vars{i})) || isstring(data.(vars{i})) || iscategorical(data.(vars{i}))
        data.(vars{i}) = grp2idx(categorical(data.(vars{i}))) - 1;
    end
end

data.Result = categorical(data.Result);

X = [data.Attendance, data.Internal_Marks, data.Assignment_Status, data.Study_Hours];
Y = data.Result;

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = fitctree(X_train, Y_train, 'SplitCriterion', 'deviance');

%% =========================
% VIEW TREE
% ==========================
view(model, 'Mode', 'graph');

%% =========================
% PREDICTION
% ==========================
Y_pred = predict(model, X_test);

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy = %.4f\n', accuracy);

classes = categories(Y);
for i = 1:length(classes)
    TP = confMat(i,i);
    FP = sum(confMat(:,i)) - TP;
    FN = sum(confMat(i,:)) - TP;

    precision = TP / (TP + FP + eps);
    recall    = TP / (TP + FN + eps);
    f1        = 2 * precision * recall / (precision + recall + eps);

    fprintf('\nClass: %s\n', string(classes(i)));
    fprintf('Precision = %.4f\n', precision);
    fprintf('Recall    = %.4f\n', recall);
    fprintf('F1-score  = %.4f\n', f1);
end

figure;
confusionchart(Y_test, Y_pred);
title('Decision Tree (ID3-style) Confusion Matrix');

File: src/data/codes/decision-tree.py
"""Decision tree — deviance ≈ log_loss criterion
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_recall_fscore_support
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(4)

if USE_SYNTHETIC:
    n = 120
    att = RNG.integers(0, 3, size=n)
    internal = RNG.integers(0, 3, size=n)
    assign = RNG.integers(0, 2, size=n)
    study = RNG.integers(0, 3, size=n)
    res = []
    for i in range(n):
        s = att[i] + internal[i] + assign[i] + study[i]
        res.append("Pass" if s >= 4 else "Fail")
    data = pd.DataFrame(
        {
            "Attendance": att,
            "Internal_Marks": internal,
            "Assignment_Status": assign,
            "Study_Hours": study,
            "Result": res,
        }
    )
else:
    data = pd.read_csv("student_performance_data.csv")

print(data.head())
data = data.dropna().drop_duplicates()

X = data[["Attendance", "Internal_Marks", "Assignment_Status", "Study_Hours"]].values
y = data["Result"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = DecisionTreeClassifier(criterion="log_loss", random_state=42).fit(X_train, y_train)
y_pred = model.predict(X_test)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print(f"\nAccuracy = {accuracy_score(y_test, y_pred):.4f}")

for lab in np.unique(y):
    m = precision_recall_fscore_support(
        y_test == lab, y_pred == lab, average="binary", zero_division=0
    )
    print(f"\nClass: {lab}")
    print(f"Precision = {m[0]:.4f}, Recall = {m[1]:.4f}, F1 = {m[2]:.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("Decision Tree (log_loss) Confusion Matrix")
plt.show()
